### DISCUSSION TITLE
osxphotos export to include face rectangles

### DISCUSSION THREAD

**Discussion by danielsbrewer** [RhetTbull/osxphotos#632] [Q&A]
Thanks for the great tool. Is it possible to use osxphotos export to put into XMP the face rectangles (either Metdata Working Group (mwg) face regions or Microsoft Photo Region Rectangle (mpri) or both). It looks like it should be possible as the FaceInfo object seems to have this information but I couldn't find an option in the "osxphotos export" command.

At the moment I am having to import everything into digikam and redo face tagging. An exiftool output of the relevant bits looks like this (for two people) after digikam:
Region Applied To Dimensions W  : 4128
Region Applied To Dimensions H  : 3096
Region Applied To Dimensions Unit: pixel
Region Name                     : Ann, Kate
Region Type                     : Face, Face
Region Area X                   : 0.336604, 0.625
Region Area Y                   : 0.444444, 0.453327
Region Area W                   : 0.119913, 0.116764
Region Area H                   : 0.186693, 0.18637
Region Area Unit                : normalized, normalized
Region Person Display Name      : Ann, Kate
Region Rectangle                : 0.276647, 0.351098, 0.119913, 0.186693, 0.566618, 0.360142, 0.116764, 0.18637

Any ideas how to get this via osxphotos?

Thanks

**Comment by RhetTbull** [comment]
osxphotos should currently do this in the XMP sidecar in both mwg and mpri.  But the sidecars have been very difficult to test as I don't have access to software that actually reads them correctly.  Here's sample output from one of the osxphotos test libraries:

```xml
<rdf:Description rdf:about=""
 xmlns:mwg-rs="http://www.metadataworkinggroup.com/schemas/regions/"
 xmlns:stArea="http://ns.adobe.com/xmp/sType/Area#"
 xmlns:stDim="http://ns.adobe.com/xap/1.0/sType/Dimensions#">
    <mwg-rs:Regions rdf:parseType="Resource">
    <mwg-rs:AppliedToDimensions rdf:parseType="Resource">
    <stDim:unit>pixel</stDim:unit>
    </mwg-rs:AppliedToDimensions>
    <mwg-rs:RegionList>
    <rdf:Bag>
     <rdf:li rdf:parseType="Resource">
      <mwg-rs:Area rdf:parseType="Resource">
      <stArea:h>0.055691</stArea:h>
      <stArea:w>0.083557</stArea:w>
      <stArea:x>0.627687</stArea:x>
      <stArea:y>0.507298</stArea:y>
      <stArea:unit>normalized</stArea:unit>
      </mwg-rs:Area>
      <mwg-rs:Name>Katie</mwg-rs:Name>
      <mwg-rs:Rotation>0.0</mwg-rs:Rotation>
      <mwg-rs:Type>Face</mwg-rs:Type>
     </rdf:li>
     <rdf:li rdf:parseType="Resource">
      <mwg-rs:Area rdf:parseType="Resource">
      <stArea:h>0.066277</stArea:h>
      <stArea:w>0.099440</stArea:w>
      <stArea:x>0.525702</stArea:x>
      <stArea:y>0.278760</stArea:y>
      <stArea:unit>normalized</stArea:unit>
      </mwg-rs:Area>
      <mwg-rs:Name></mwg-rs:Name>
      <mwg-rs:Rotation>0.0</mwg-rs:Rotation>
      <mwg-rs:Type>Face</mwg-rs:Type>
     </rdf:li>
     <rdf:li rdf:parseType="Resource">
      <mwg-rs:Area rdf:parseType="Resource">
      <stArea:h>0.064400</stArea:h>
      <stArea:w>0.096624</stArea:w>
      <stArea:x>0.349200</stArea:x>
      <stArea:y>0.384542</stArea:y>
      <stArea:unit>normalized</stArea:unit>
      </mwg-rs:Area>
      <mwg-rs:Name>Suzy</mwg-rs:Name>
      <mwg-rs:Rotation>0.0</mwg-rs:Rotation>
      <mwg-rs:Type>Face</mwg-rs:Type>
     </rdf:li>
    </rdf:Bag>
    </mwg-rs:RegionList>
    </mwg-rs:Regions>
</rdf:Description>
<rdf:Description rdf:about=""
 xmlns:MP="http://ns.microsoft.com/photo/1.2/"
 xmlns:MPRI="http://ns.microsoft.com/photo/1.2/t/RegionInfo#"
 xmlns:MPReg="http://ns.microsoft.com/photo/1.2/t/Region#">
    <MP:RegionInfo rdf:parseType="Resource">
     <MPRI:Regions>
      <rdf:Bag>
       <rdf:li rdf:parseType="Resource">
       <MPReg:PersonDisplayName>Katie</MPReg:PersonDisplayName>
       <MPReg:Rectangle>0.585908, 0.479452, 0.083557, 0.055691</MPReg:Rectangle>
       </rdf:li>
       <rdf:li rdf:parseType="Resource">
       <MPReg:PersonDisplayName></MPReg:PersonDisplayName>
       <MPReg:Rectangle>0.475982, 0.245622, 0.099440, 0.066277</MPReg:Rectangle>
       </rdf:li>
       <rdf:li rdf:parseType="Resource">
       <MPReg:PersonDisplayName>Suzy</MPReg:PersonDisplayName>
       <MPReg:Rectangle>0.300888, 0.352342, 0.096624, 0.064400</MPReg:Rectangle>
       </rdf:li>
      </rdf:Bag>
     </MPRI:Regions>
    </MP:RegionInfo>
</rdf:Description>
```

If you can provide an XMP with face regions properly coded that DigiKam will read I can take a look at what I'm doing wrong.

**Reply by RhetTbull** [reply]
Looks like it doesn't have the applied to dimensions, just the units.

**Comment by RhetTbull** [comment]
Could you provide the exiftool output in json format using `exiftool -j -G file.jpg` for a photo with face regions.  I could then use that to also add this to the exiftool sidecar.

**Reply by danielsbrewer** [reply]
Thanks for looking into this. Note I don't use XMP sidecars but in the actual image file. Here is the output:

[{
  "SourceFile": "20171020_174722.jpg",
  "ExifTool:ExifToolVersion": 12.30,
  "ExifTool:Warning": "[minor] Fixed incorrect URI for xmlns:MicrosoftPhoto",
  "File:FileName": "20171020_174722.jpg",
  "File:Directory": ".",
  "File:FileSize": "1004 KiB",
  "File:FileModifyDate": "2022:02:11 13:08:53+00:00",
  "File:FileAccessDate": "2022:02:11 13:09:03+00:00",
  "File:FileInodeChangeDate": "2022:02:11 13:09:01+00:00",
  "File:FilePermissions": "-rw-r--r--",
  "File:FileType": "JPEG",
  "File:FileTypeExtension": "jpg",
  "File:MIMEType": "image/jpeg",
  "File:ExifByteOrder": "Little-endian (Intel, II)",
  "File:CurrentIPTCDigest": "2e86517aac4b8574bc852223de074cf0",
  "File:ImageWidth": 4128,
  "File:ImageHeight": 3096,
  "File:EncodingProcess": "Baseline DCT, Huffman coding",
  "File:BitsPerSample": 8,
  "File:ColorComponents": 3,
  "File:YCbCrSubSampling": "YCbCr4:2:0 (2 2)",
  "JFIF:JFIFVersion": 1.01,
  "JFIF:ResolutionUnit": "None",
  "JFIF:XResolution": 1,
  "JFIF:YResolution": 1,
  "EXIF:ImageWidth": 4128,
  "EXIF:ImageHeight": 3096,
  "EXIF:DocumentName": "Granny's seventieth",
  "EXIF:Make": "samsung",
  "EXIF:Model": "SM-J510FN",
  "EXIF:Orientation": "Horizontal (normal)",
  "EXIF:XResolution": 72,
  "EXIF:YResolution": 72,
  "EXIF:ResolutionUnit": "inches",
  "EXIF:Software": "J510FNXXU2AQG3",
  "EXIF:ModifyDate": "2017:10:20 17:47:21",
  "EXIF:YCbCrSubSampling": "YCbCr4:2:0 (2 2)",
  "EXIF:YCbCrPositioning": "Centered",
  "EXIF:ExposureTime": "1/17",
  "EXIF:FNumber": 1.9,
  "EXIF:ExposureProgram": "Program AE",
  "EXIF:ISO": 1250,
  "EXIF:ExifVersion": "0220",
  "EXIF:DateTimeOriginal": "2017:10:20 17:47:21",
  "EXIF:CreateDate": "2017:10:20 17:47:21",
  "EXIF:OffsetTimeOriginal": "+01:00",
  "EXIF:ComponentsConfiguration": "Y, Cb, Cr, -",
  "EXIF:ShutterSpeedValue": "1/17",
  "EXIF:ApertureValue": 1.9,
  "EXIF:BrightnessValue": -3.03,
  "EXIF:ExposureCompensation": 0,
  "EXIF:MaxApertureValue": 1.9,
  "EXIF:MeteringMode": "Center-weighted average",
  "EXIF:LightSource": "Unknown",
  "EXIF:Flash": "No Flash",
  "EXIF:FocalLength": "3.7 mm",
  "EXIF:UserComment": "\n",
  "EXIF:FlashpixVersion": "0100",
  "EXIF:ColorSpace": "sRGB",
  "EXIF:ExifImageWidth": 4128,
  "EXIF:ExifImageHeight": 3096,
  "EXIF:InteropIndex": "R98 - DCF basic file (sRGB)",
  "EXIF:InteropVersion": "0100",
  "EXIF:SensingMethod": "One-chip color area",
  "EXIF:SceneType": "Directly photographed",
  "EXIF:ExposureMode": "Auto",
  "EXIF:WhiteBalance": "Auto",
  "EXIF:FocalLengthIn35mmFormat": "28 mm",
  "EXIF:SceneCaptureType": "Standard",
  "EXIF:ImageUniqueID": "44092c40dbe96b1e0000000000000000",
  "EXIF:GPSVersionID": "2.3.0.0",
  "EXIF:GPSLatitudeRef": "North",
  "EXIF:GPSLatitude": "52 deg 14' 18.15\"",
  "EXIF:GPSLongitudeRef": "West",
  "EXIF:GPSLongitude": "0 deg 47' 2.32\"",
  "EXIF:GPSAltitudeRef": "Above Sea Level",
  "EXIF:GPSAltitude": "0 m",
  "EXIF:GPSMapDatum": "WGS-84",
  "EXIF:XPTitle": "Granny's seventieth",
  "EXIF:XPKeywords": "Ann Ogden, Kate Ogden, Oct 2017",
  "EXIF:Compression": "JPEG (old-style)",
  "EXIF:ThumbnailOffset": 3586,
  "EXIF:ThumbnailLength": 6363,
  "EXIF:ThumbnailImage": "(Binary data 6363 bytes, use -b option to extract)",
  "XMP:XMPToolkit": "XMP Core 5.5.0",
  "XMP:GPSAltitude": "0 m",
  "XMP:GPSAltitudeRef": "Above Sea Level",
  "XMP:GPSLatitude": "52 deg 14' 18.15\" N",
  "XMP:GPSLongitude": "0 deg 47' 2.32\" W",
  "XMP:GPSMapDatum": "WGS-84",
  "XMP:City": "Cogenhoe",
  "XMP:Country": "United Kingdom",
  "XMP:State": "England",
  "XMP:MetadataDate": "2019:03:22 19:56:58+00:00",
  "XMP:Categories": "<Categories><Category Assigned=\"0\">People<Category Assigned=\"1\">Ann Ogden</Category></Category><Category Assigned=\"1\">Oct 2017</Category><Category Assigned=\"1\">Kate Ogden</Category></Categories>",
  "XMP:LocationCreatedCity": "Cogenhoe",
  "XMP:LocationCreatedCountryName": "United Kingdom",
  "XMP:LocationCreatedProvinceState": "England",
  "XMP:PersonInImage": ["Ann Ogden","Kate Ogden"],
  "XMP:Description": "Granny's seventieth",
  "XMP:Subject": ["Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","Kate Ogden"],
  "XMP:Title": "Granny's seventieth",
  "XMP:TagsList": ["People/Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","People/Kate Ogden"],
  "XMP:LastKeywordXMP": ["People/Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","People/Kate Ogden"],
  "XMP:HierarchicalSubject": ["People|Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","People|Kate Ogden"],
  "XMP:CatalogSets": ["People|Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","People|Kate Ogden"],
  "XMP:RegionAppliedToDimensionsW": 4128,
  "XMP:RegionAppliedToDimensionsH": 3096,
  "XMP:RegionAppliedToDimensionsUnit": "pixel",
  "XMP:RegionName": ["Ann Ogden","Kate Ogden"],
  "XMP:RegionType": ["Face","Face"],
  "XMP:RegionAreaX": [0.336604,0.625],
  "XMP:RegionAreaY": [0.444444,0.453327],
  "XMP:RegionAreaW": [0.119913,0.116764],
  "XMP:RegionAreaH": [0.186693,0.18637],
  "XMP:RegionAreaUnit": ["normalized","normalized"],
  "XMP:RegionPersonDisplayName": ["Ann Ogden","Kate Ogden"],
  "XMP:RegionRectangle": ["0.276647, 0.351098, 0.119913, 0.186693","0.566618, 0.360142, 0.116764, 0.18637"],
  "IPTC:EnvelopeRecordVersion": 4,
  "IPTC:CodedCharacterSet": "UTF8",
  "IPTC:ApplicationRecordVersion": 4,
  "IPTC:ObjectName": "Granny's seventieth",
  "IPTC:Keywords": ["Ann Ogden","Oct 2017","Ann Ogden","Kate Ogden","Kate Ogden"],
  "IPTC:DateCreated": "2017:10:20",
  "IPTC:TimeCreated": "17:47:21+00:00",
  "IPTC:City": "Cogenhoe",
  "IPTC:Province-State": "England",
  "IPTC:Country-PrimaryLocationName": "United Kingdom",
  "IPTC:Headline": "Granny's seventieth",
  "Photoshop:IPTCDigest": "2e86517aac4b8574bc852223de074cf0",
  "Composite:Aperture": 1.9,
  "Composite:ImageSize": "4128x3096",
  "Composite:Megapixels": 12.8,
  "Composite:ScaleFactor35efl": 7.6,
  "Composite:ShutterSpeed": "1/17",
  "Composite:SubSecDateTimeOriginal": "2017:10:20 17:47:21+01:00",
  "Composite:GPSAltitude": "0 m Above Sea Level",
  "Composite:GPSLatitude": "52 deg 14' 18.15\" N",
  "Composite:GPSLongitude": "0 deg 47' 2.32\" W",
  "Composite:DateTimeCreated": "2017:10:20 17:47:21+00:00",
  "Composite:GPSLatitudeRef": "North",
  "Composite:GPSLongitudeRef": "West",
  "Composite:CircleOfConfusion": "0.004 mm",
  "Composite:FOV": "65.5 deg",
  "Composite:FocalLength35efl": "3.7 mm (35 mm equivalent: 28.0 mm)",
  "Composite:GPSPosition": "52 deg 14' 18.15\" N, 0 deg 47' 2.32\" W",
  "Composite:HyperfocalDistance": "1.81 m",
  "Composite:LightValue": 2.3
}]

**Reply by RhetTbull** [reply]
Thanks -- that should be enough for me to get this implemented.  How are you exporting the data from osxphotos for import into DigiKam?  (e.g. XMP sidecar, exiftool sidecar, --exiftool?)

**Reply by danielsbrewer** [reply]
This is the command line tha I use. I export album by album.

osxphotos export --album "September 2019" --touch-file --skip-original-if-edited --exiftool  --exiftool-merge-keywords  --exiftool-merge-persons --person-keyword --album-keyword --directory '{folder_album( - )}'

**Reply by danielsbrewer** [reply]
Here is a link to the image after digikam.

**Reply by danielsbrewer** [reply]
https://ueanorwich-my.sharepoint.com/:i:/g/personal/nsu13cbu_uea_ac_uk/EYpkuI72bZ1Ov6cd3zHX5MsB8IpM_HAD8AoQ9ZlldVxLlg?e=j9VFNV

**Reply by RhetTbull** [reply]
Perfect -- thanks!  That will be helpful for debugging.

**Reply by RhetTbull** [reply]
@danielsbrewer I've implemented a beta fix for this in [v0.45.11](https://github.com/RhetTbull/osxphotos/releases/tag/v0.45.11).  Upgrade and give it a try.  You'll need to add the `--beta` flag to your command line to enable the fix.  Once I know it's working I'll enable it by default in the next release.  

I tried in digikam and it is picking up the face regions.

<img width="596" alt="Screen Shot 2022-02-13 at 12 29 08 AM" src="https://user-images.githubusercontent.com/41546558/153745270-139e87e5-51c2-4c3f-9586-27d847347945.png">

**Reply by danielsbrewer** [reply]
I have tried it on a whole album of photos and it works really really well. Thank you. That is absolutely amazing. Both the rectangles and name tags come up clearly in digiKam.

The only difference I saw was that in the output from exiftool that it didn't have this bit (that was created by digiKam):
"XMP:RegionPersonDisplayName": ["Ann","Kate"],
"XMP:RegionRectangle": ["0.276647, 0.351098, 0.119913, 0.186693","0.566618, 0.360142, 0.116764, 0.18637"],

But for my purposes this isn't an issue (and seems a duplication of the information) but I thought I would let you know. Thanks again.

**Reply by RhetTbull** [reply]
> The only difference I saw was that in the output from exiftool that it didn't have this bit (that was created by digiKam):
>"XMP:RegionPersonDisplayName": ["Ann","Kate"],
>"XMP:RegionRectangle": ["0.276647, 0.351098, 0.119913, 0.186693","0.566618, 0.360142, 0.116764, 0.18637"],

Yes, I couldn't find a reference for how to compute the region coordinates so left it out. I tried to make some educated guesses (which were wrong) and ended up with two rectangles in digikam for each face (so I surmised that this information was duplicative). I suppose I could look at the digikam source to figure it out but as it appears to be working, I have other bugs to squash.

**Reply by danielsbrewer** [reply]
That sounds reasonable to me. Thanks again.

**Comment by RhetTbull** [comment]
I've enabled this by default in 0.45.12 so you won't need the --beta flag once upgraded.
